### PR TITLE
feat(EntityService): batched transactions and ebean updates

### PR DESCRIPTION
**Pull request by david-leifker**
* Removes MAE producer code and logic around MAE in EntityService
* Adds batching of database aspect operations in EntityService
* Exposes the Async futures (and handles them) when producing to MCL topic
* Update ebean library
* Fix dropped writes with multi-threading when deadlocked
* Multithreading test added, along with fake data generator
* Only generate default aspects for UPSERT/CREATE change types

## Checklist

- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

File: metadata-io/src/main/java/com/linkedin/metadata/entity/EntityServiceImpl.java [hunk 1]
```diff
@@ -527,227 +481,192 @@ public ListResult<RecordTemplate> listLatestAspects(
...
+      // read #2
+      Map<String, Map<String, Long>> nextVersions = _aspectDao.getNextVersions(urnAspects);
+
+      List<UpsertBatchItem> items = aspectsBatch.getItems().stream()
+              .map(item -> {
+                if (item instanceof UpsertBatchItem) {
+                  return (UpsertBatchItem) item;
+                } else {
+                  // patch to upsert
+                  PatchBatchItem patchBatchItem = (PatchBatchItem) item;
```
Comment (david-leifker) [inline metadata-io/src/main/java/com/linkedin/metadata/entity/EntityServiceImpl.java:564]: Merged the paths for PATCH and Upsert, they were practically identical except we take a patch and apply it to the latest aspect. Essentially patching just converts it from a patch item to a normal upsert item.

File: metadata-io/src/main/java/com/linkedin/metadata/entity/EntityServiceImpl.java [hunk 2]
```diff
@@ -527,227 +481,192 @@ public ListResult<RecordTemplate> listLatestAspects(
...
+                          .maxVersion(latest.getVersion())
+                          .build();
+                }
+
+                return result;
+              }).collect(Collectors.toList());
+
+      // commit upserts prior to retention or kafka send, if supported by impl
+      if (tx != null) {
+        tx.commitAndContinue();
```
Comment (david-leifker) [inline metadata-io/src/main/java/com/linkedin/metadata/entity/EntityServiceImpl.java:611]: After the aspect insert, we commit before proceeding to the possible application of retention policy. If the retention logic fails, at least we still get the upserts of the aspects.

File: metadata-io/src/main/java/com/linkedin/metadata/entity/EntityServiceImpl.java [hunk 3]
```diff
@@ -527,227 +481,192 @@ public ListResult<RecordTemplate> listLatestAspects(
...
+              }
+            }).collect(Collectors.toList());
+
+    if (statusPairs.stream().anyMatch(p -> !p.getFirst())) {
+      log.error("Failed to produce MCLs: {}", statusPairs.stream()
+              .filter(p -> !p.getFirst())
+              .map(Pair::getValue)
+              .map(v -> v.getRequest().toString())
+              .collect(Collectors.toList()));
+      // TODO restoreIndices?
```
Comment (david-leifker) [inline metadata-io/src/main/java/com/linkedin/metadata/entity/EntityServiceImpl.java:665]: This is the case where the SQL commit was successful, otherwise we wouldn't be emitting MCLs. In this case one or more MCLs fail to be produced. We might try to recover this by running restoreIndices automatically, however it would also depend on producing messages to kafka and we'd end up in a loop. We are exposing this as a failure to the client, but they would have to call restoreIndices rather then retry to see any effect.

File: metadata-io/src/main/java/com/linkedin/metadata/entity/EntityServiceImpl.java [hunk 4]
```diff
@@ -1054,43 +871,19 @@ public String batchApplyRetention(Integer start, Integer count, Integer attemptW
...
-
-  private void preprocessEvent(MetadataChangeLog metadataChangeLog) {
+  private boolean preprocessEvent(MetadataChangeLog metadataChangeLog) {
     if (_preProcessHooks.isUiEnabled()) {
       if (metadataChangeLog.getSystemMetadata() != null) {
         if (metadataChangeLog.getSystemMetadata().getProperties() != null) {
           if (UI_SOURCE.equals(metadataChangeLog.getSystemMetadata().getProperties().get(APP_SOURCE))) {
             // Pre-process the update indices hook for UI updates to avoid perceived lag from Kafka
             _updateIndicesService.handleChangeEvent(metadataChangeLog);
+            return true;
```
Comment (david-leifker) [inline metadata-io/src/main/java/com/linkedin/metadata/entity/EntityServiceImpl.java:881]: Lets report back up the result chain whether we actually ran this. More of an fyi at this point.

File: metadata-io/src/main/java/com/linkedin/metadata/entity/EntityServiceImpl.java [hunk 5]
```diff
@@ -1293,60 +1061,59 @@ public Map<Urn, Entity> getEntities(@Nonnull final Set<Urn> urns, @Nonnull Set<S
...
-    } else {
-      final Snapshot newSnapshot = buildSnapshot(urn, newAspectValue);
-      Snapshot oldSnapshot = null;
-      if (oldAspectValue != null) {
-        oldSnapshot = buildSnapshot(urn, oldAspectValue);
-      }
-      _producer.produceMetadataAuditEvent(urn, oldSnapshot, newSnapshot, oldSystemMetadata, newSystemMetadata,
-          operation);
-    }
+  public Pair<Future<?>, Boolean> alwaysProduceMCLAsync(@Nonnull final Urn urn, @Nonnull final AspectSpec aspectSpec,
```
Comment (david-leifker) [inline metadata-io/src/main/java/com/linkedin/metadata/entity/EntityServiceImpl.java:1064]: Introducing some naming here, 2 key points. We have another path which will conditionally produce MCLs based on the aspect, flags, etc. Lets be clear in that using this doesn't follow that path. Second, and probably more important is the communication to the caller that these are *Async*. There is a future in the from the kafka producer that should be used and not dropped.

File: .github/workflows/docker-unified.yml [hunk 6]
```diff
@@ -429,8 +429,8 @@ jobs:
         kafka_setup_build,
         mysql_setup_build,
         elasticsearch_setup_build,
-        mae_consumer_build,
-        mce_consumer_build,
+#        mae_consumer_build,
```
Comment (david-leifker) [inline .github/workflows/docker-unified.yml:432]: Temporary until space issues can be resolved.

**Comment by david-leifker** [comment]
I want to add a test, ebean in-memory, to monitor ingestion queries for potentially duplicate queries for a given set of test data used in the smoke tests.

File: metadata-io/src/test/java/com/linkedin/metadata/AspectUtilsTest.java [hunk 7]
```diff
@@ -54,7 +54,7 @@ public void testAdditionalChanges() {
     DatasetProperties datasetProperties = new DatasetProperties().setName("name");
     proposal1.setAspect(GenericRecordUtils.serializeAspect(datasetProperties));
     proposal1.setEntityType("dataset");
-    proposal1.setChangeType(ChangeType.PATCH);
+    proposal1.setChangeType(ChangeType.UPSERT);
```
Comment (david-leifker) [inline metadata-io/src/test/java/com/linkedin/metadata/AspectUtilsTest.java:57]: @RyanHolstien - Reminder to discuss this. I do not think that we should be generating additional default aspects because I believe that PATCH should only apply to existing entities. If they exist, then these should have been generated already.

**Comment by david-leifker** [comment]
Create flag for : Only generate default aspects for UPSERT/CREATE change types

File: metadata-io/src/test/java/com/linkedin/metadata/AspectUtilsTest.java (hunk 7 above)
Comment (david-leifker) [reply metadata-io/src/test/java/com/linkedin/metadata/AspectUtilsTest.java:57]: going to revert this

**Review by RyanHolstien — APPROVED** [review]
